libethereum/BlockChainSync.cpp

```diff
@@ -463,6 +463,11 @@ void BlockChainSync::onPeerBlockHeaders(std::shared_ptr<EthereumPeer> _peer, RLP
 					m_downloadingHeaders.erase(blockNumber - 1);
 					removeItem(m_headers, blockNumber - 1);
 					removeItem(m_bodies, blockNumber - 1);
+					m_lastImportedBlock--;
+					m_highestBlock = m_lastImportedBlock;
```

Comment:
Right, better do that under the `blockNumber == m_lastImportedBlock + 1 && info.parentHash() != m_lastImportedBlockHash` only and get `m_lastImportedBlockHash` from the block chain or block queue. Better option would be to restart sync here and set `m_haveCommonBlock` to false.
Also resetting `m_highestBlock` does not make much sense. It is used for progress reporting only.